File: src/readers/cursor/authState.ts
import { existsSync } from 'node:fs';
import { copyFile, mkdtemp, rm } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { DatabaseSync } from 'node:sqlite';

export interface CursorAuthState {
  accessToken?: string;
  refreshToken?: string;
}

function normalizeCursorDatabaseValue(value: unknown): string | undefined {
  if (typeof value === 'string') {
    const trimmed = value.trim();
    return trimmed === '' ? undefined : trimmed;
  }
  if (value instanceof Uint8Array) {
    const trimmed = Buffer.from(value).toString('utf8').trim();
    return trimmed === '' ? undefined : trimmed;
  }
  return undefined;
}

function readCursorAuthStateFromDatabase(databasePath: string): CursorAuthState {
  const database = new DatabaseSync(databasePath, { readOnly: true });
  try {
    const query = database.prepare('SELECT value FROM ItemTable WHERE key = ? LIMIT 1');
    const accessRow = query.get('cursorAuth/accessToken') as { value?: unknown } | undefined;
    const refreshRow = query.get('cursorAuth/refreshToken') as { value?: unknown } | undefined;
    return {
      accessToken: normalizeCursorDatabaseValue(accessRow?.value),
      refreshToken: normalizeCursorDatabaseValue(refreshRow?.value),
    };
  } finally {
    database.close();
  }
}

function isSqliteLockedError(error: unknown): boolean {
  return error instanceof Error && /database is locked/iu.test(error.message);
}

async function withCursorStateSnapshot<T>(
  databasePath: string,
  callback: (snapshotPath: string) => T | Promise<T>,
): Promise<T> {
  const snapshotDir = await mkdtemp(join(tmpdir(), 'aitrack-cursor-'));
  const snapshotPath = join(snapshotDir, 'state.vscdb');
  await copyFile(databasePath, snapshotPath);
  for (const suffix of ['-shm', '-wal']) {
    const companionPath = `${databasePath}${suffix}`;
    if (!existsSync(companionPath)) continue;
    await copyFile(companionPath, `${snapshotPath}${suffix}`);
  }
  try {
    return await callback(snapshotPath);
  } finally {
    await rm(snapshotDir, { recursive: true, force: true });
  }
}

export async function readCursorAuthState(databasePath: string): Promise<CursorAuthState> {
  try {
    return readCursorAuthStateFromDatabase(databasePath);
  } catch (error) {
    if (!isSqliteLockedError(error)) throw error;
    const state = await withCursorStateSnapshot(databasePath, (snapshotPath) =>
      readCursorAuthStateFromDatabase(snapshotPath),
    );
    return state;
  }
}

File: crates/aitrack-core/src/readers/cursor/auth_state.rs
//! Reading Cursor auth credentials from state.vscdb.

use rusqlite::{Connection, OpenFlags};
use std::error::Error;
use std::fs;
use std::path::Path;

#[derive(Debug, Clone)]
pub struct CursorAuthState {
    pub access_token: Option<String>,
    pub refresh_token: Option<String>,
}

fn normalize_cursor_database_value(value: Option<Vec<u8>>) -> Option<String> {
    value.and_then(|bytes| {
        let s = String::from_utf8_lossy(&bytes);
        let trimmed = s.trim();
        if trimmed.is_empty() {
            None
        } else {
            Some(trimmed.to_string())
        }
    })
}

fn read_cursor_auth_state_from_database(database_path: &Path) -> anyhow::Result<CursorAuthState> {
    let conn = Connection::open_with_flags(database_path, OpenFlags::SQLITE_OPEN_READ_ONLY)?;

    let access_token = conn
        .query_row(
            "SELECT value FROM ItemTable WHERE key = ? LIMIT 1",
            ["cursorAuth/accessToken"],
            |row| row.get::<_, Vec<u8>>(0),
        )
        .ok();

    let refresh_token = conn
        .query_row(
            "SELECT value FROM ItemTable WHERE key = ? LIMIT 1",
            ["cursorAuth/refreshToken"],
            |row| row.get::<_, Vec<u8>>(0),
        )
        .ok();

    Ok(CursorAuthState {
        access_token: normalize_cursor_database_value(access_token),
        refresh_token: normalize_cursor_database_value(refresh_token),
    })
}

fn is_sqlite_locked_error(error: &dyn Error) -> bool {
    error
        .to_string()
        .to_lowercase()
        .contains("database is locked")
}

async fn with_cursor_state_snapshot<T, F>(database_path: &Path, callback: F) -> anyhow::Result<T>
where
    F: FnOnce(&Path) -> anyhow::Result<T>,
{
    let temp_dir = tempfile::Builder::new()
        .prefix("aitrack-cursor-")
        .tempdir()?;
    let snapshot_path = temp_dir.path().join("state.vscdb");

    fs::copy(database_path, &snapshot_path)?;

    // Copy WAL and SHM files if they exist
    for suffix in &["-shm", "-wal"] {
        let companion_path = format!("{}{}", database_path.display(), suffix);
        if Path::new(&companion_path).exists() {
            let snapshot_companion = format!("{}{}", snapshot_path.display(), suffix);
            let _ = fs::copy(&companion_path, &snapshot_companion);
        }
    }

    callback(&snapshot_path)
}

pub async fn read_cursor_auth_state(database_path: &Path) -> anyhow::Result<CursorAuthState> {
    match read_cursor_auth_state_from_database(database_path) {
        Ok(state) => Ok(state),
        Err(err) if is_sqlite_locked_error(err.as_ref()) => {
            with_cursor_state_snapshot(database_path, |snapshot_path| {
                read_cursor_auth_state_from_database(snapshot_path)
            })
            .await
        }
        Err(err) => Err(err),
    }
}
